In [19]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [20]:
hbd = pd.read_csv('hotel_bookings_data.csv')
hbd.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2017          September   
1  Resort Hotel            0        737               2017          September   
2  Resort Hotel            0          7               2017          September   
3  Resort Hotel            0         13               2017          September   
4  Resort Hotel            0         14               2017          September   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_weekdays_nights  adults  ...  \
0                        0                         0       2  ...   
1                        0                         0       2  ...   
2                        0                         1       1  ...   
3                        0                         1       1  ...   
4                        0                         2       2  ...   

   booking_changes  deposit_type  agent company days_in_waiting_list  \
0                3    No Deposit    NaN     NaN                    0   
1                4    No Deposit    NaN     NaN                    0   
2                0    No Deposit    NaN     NaN                    0   
3                0    No Deposit  304.0     NaN                    0   
4                0    No Deposit  240.0     NaN                    0   

  customer_type   adr  required_car_parking_spaces  total_of_special_requests  \
0      Personal   0.0                            0                          0   
1      Personal   0.0                            0                          0   
2      Personal  75.0                            0                          0   
3      Personal  75.0                            0                          0   
4      Personal  98.0                            0                          1   

   reservation_status  
0           Check-Out  
1           Check-Out  
2           Check-Out  
3           Check-Out  
4           Check-Out  

[5 rows x 29 columns]

In [21]:
hbd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_weekdays_nights        119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [22]:
list_item = []
for col in hbd.columns :
    list_item.append([col, hbd[col].dtype, hbd[col].isna().sum(), 100*hbd[col].isna().sum()/len(hbd[col]), hbd[col].nunique(), hbd[col].unique()[:5]])
desc_hbd = pd.DataFrame(data=list_item, columns = 'Feature DataType Null NullPct UniqueNum UniqueSample'.split())
desc_hbd

Feature DataType    Null    NullPct  UniqueNum  \
0                            hotel   object       0   0.000000          2   
1                      is_canceled    int64       0   0.000000          2   
2                        lead_time    int64       0   0.000000        479   
3                arrival_date_year    int64       0   0.000000          3   
4               arrival_date_month   object       0   0.000000         12   
5         arrival_date_week_number    int64       0   0.000000         53   
6        arrival_date_day_of_month    int64       0   0.000000         31   
7          stays_in_weekend_nights    int64       0   0.000000         17   
8         stays_in_weekdays_nights    int64       0   0.000000         35   
9                           adults    int64       0   0.000000         14   
10                        children  float64       4   0.003350          5   
11                          babies    int64       0   0.000000          5   
12                            meal   object       0   0.000000          5   
13                            city   object     488   0.408744        177   
14                  market_segment   object       0   0.000000          8   
15            distribution_channel   object       0   0.000000          5   
16               is_repeated_guest    int64       0   0.000000          2   
17          previous_cancellations    int64       0   0.000000         15   
18  previous_bookings_not_canceled    int64       0   0.000000         73   
19                 booking_changes    int64       0   0.000000         21   
20                    deposit_type   object       0   0.000000          3   
21                           agent  float64   16340  13.686238        333   
22                         company  float64  112593  94.306893        352   
23            days_in_waiting_list    int64       0   0.000000        128   
24                   customer_type   object       0   0.000000          4   
25                             adr  float64       0   0.000000       8879   
26     required_car_parking_spaces    int64       0   0.000000          5   
27       total_of_special_requests    int64       0   0.000000          6   
28              reservation_status   object       0   0.000000          3   

                                         UniqueSample  
0                          [Resort Hotel, City Hotel]  
1                                              [0, 1]  
2                               [342, 737, 7, 13, 14]  
3                                  [2017, 2018, 2019]  
4   [September, October, November, December, January]  
5                                [27, 28, 29, 30, 31]  
6                                     [1, 2, 3, 4, 5]  
7                                     [0, 1, 2, 4, 3]  
8                                     [0, 1, 2, 3, 4]  
9                                    [2, 1, 3, 4, 40]  
10                         [0.0, 1.0, 2.0, 10.0, 3.0]  
11                                   [0, 1, 2, 10, 9]  
12  [Breakfast, Full Board, Dinner, No Meal, Undef...  
13  [Kota Denpasar, Kabupaten Bangka, Kabupaten Sl...  
14  [Direct, Corporate, Online TA, Offline TA/TO, ...  
15         [Direct, Corporate, TA/TO, Undefined, GDS]  
16                                             [0, 1]  
17                                   [0, 1, 2, 3, 26]  
18                                    [0, 1, 2, 3, 4]  
19                                    [3, 4, 0, 1, 2]  
20               [No Deposit, Refundable, Non Refund]  
21                   [nan, 304.0, 240.0, 303.0, 15.0]  
22                  [nan, 110.0, 113.0, 270.0, 178.0]  
23                               [0, 50, 47, 65, 122]  
24            [Personal, Contract, Family, Bussiness]  
25                    [0.0, 75.0, 98.0, 107.0, 103.0]  
26                                    [0, 1, 2, 8, 3]  
27                                    [0, 1, 3, 2, 4]  
28                     [Check-Out, Canceled, No-Show]

In [23]:
hbd.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2018.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2017.000000   
25%         0.000000      18.000000        2018.000000   
50%         0.000000      69.000000        2018.000000   
75%         1.000000     160.000000        2019.000000   
max         1.000000     737.000000        2019.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_weekdays_nights         adults  \
count            119390.000000             119390.000000  119390.000000   
mean                  0.927599                  2.500302       1.856403   
std                   0.998613                  1.908286       0.579261   
min                   0.000000                  0.000000       0.000000   
25%                   0.000000                  1.000000       2.000000   
50%                   1.000000                  2.000000       2.000000   
75%                   2.000000                  3.000000       2.000000   
max                  19.000000                 50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000             

In [24]:
hbd.duplicated().sum()

33261

In [25]:
# Create new column 'total_guests' from the combination of 'adults', 'children', 'babies' columns
hbd['total_guests'] = hbd['adults'] + hbd['children'] + hbd['babies']
print('Data with 0 guest: {0} out of {1} all data'.format(hbd[hbd['total_guests'] == 0].shape[0], hbd.shape[0]))
# Create new column 'stay_duration' from the combination of 'stays_in_weekend_nights' and 'stays_in_weekdays_nights'
hbd['stay_duration'] = hbd['stays_in_weekend_nights'] + hbd['stays_in_weekdays_nights']
print('Data with 0 night: {0} out of {1} all data'.format(hbd[hbd['stay_duration'] == 0].shape[0], hbd.shape[0]))

Data with 0 guest: 180 out of 119390 all data
Data with 0 night: 715 out of 119390 all data


Obzervation :
* There are 119390 rows with 29 features the target feature is reservation_status
* There are nulls in children, city, agent and company features
* There are value called Undefined in meal feature
* There are 33261 duplicated rows
* There are unecessary values with 0 guests and 0 nights
* There are unecessary data type

Solution to these problems are :
* Drop 33261 duplicated rows
* Performing imputation for children, city, agent, and company features
* Change value Undifined to No Meal in meal feature
* Drop values with condition 0 guests and 0 nights
* Change datatype from float to integer

# Data Preprocessing

In [26]:
#Drop duplicated data
hbd.drop_duplicates(inplace=True)
print ('Total duplicated data after dropping:', hbd.duplicated().sum())

Total duplicated data after dropping: 0


In [27]:
# Handling null values
# Null values in 'children' feature could mean 0 children
hbd['children'] = hbd['children'].fillna(0)
# Null values in 'city' feature can be seen as Unknown city
hbd['city'] = hbd['city'].fillna('Unknown')
# According to the research paper attached to the dataset, null values in 'agent' and 'company' means the booking did not come from a travel agent. So, it can be replaced by 0, meaning 'Not applicable'. 
hbd['agent'] = hbd['agent'].fillna(0)
hbd['company'] = hbd['company'].fillna(0)
# Check null values again
hbd.isna().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_weekdays_nights          0
adults                            0
children                          0
babies                            0
meal                              0
city                              0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests   

In [28]:
# Fix the data type
hbd['children'] = hbd['children'].astype('int64')
hbd['agent'] = hbd['agent'].astype('int64')
hbd['company'] = hbd['company'].astype('int64')

In [29]:
# Fix incorrect value
# Replace 'Undefined' in 'meal' feature with 'No Meal'
hbd['meal'] = hbd['meal'].replace(['Undefined'], 'No Meal')
hbd['meal'].value_counts()

Breakfast     67088
No Meal        9896
Dinner         8798
Full Board      347
Name: meal, dtype: int64

In [30]:
# Remove unnecessary data with 0 guests and 0 nights
hbdc = hbd[(hbd['total_guests'] > 0) & (hbd['stay_duration'] > 0)].copy()

print('before pre-processing:', hbd.shape[0])
print('after pre-processing:', hbdc.shape[0])

before pre-processing: 86129
after pre-processing: 85374


In [31]:
hbd.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 86129 entries, 0 to 119389
Data columns (total 31 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           86129 non-null  object 
 1   is_canceled                     86129 non-null  int64  
 2   lead_time                       86129 non-null  int64  
 3   arrival_date_year               86129 non-null  int64  
 4   arrival_date_month              86129 non-null  object 
 5   arrival_date_week_number        86129 non-null  int64  
 6   arrival_date_day_of_month       86129 non-null  int64  
 7   stays_in_weekend_nights         86129 non-null  int64  
 8   stays_in_weekdays_nights        86129 non-null  int64  
 9   adults                          86129 non-null  int64  
 10  children                        86129 non-null  int64  
 11  babies                          86129 non-null  int64  
 12  meal                           